client/render/AssetGenerator.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * AssetGenerator.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#include "StdInc.h"
+#include "AssetGenerator.h"
+
+#include "../gui/CGuiHandler.h"
+#include "../render/IImage.h"
+#include "../render/IImageLoader.h"
+#include "../render/Canvas.h"
+#include "../render/IRenderHandler.h"
+
+#include "../lib/filesystem/Filesystem.h"
+
+void AssetGenerator::generate()
+{
+	createBigSpellBook();
+	createAdventureOptionsCleanBackground();
+}
+
+void AssetGenerator::createAdventureOptionsCleanBackground()
+{
+	std::string filename = "data/ADVOPTBC.bmp";
+
+	if(CResourceHandler::get()->existsResource(ResourcePath(filename))) // overridden by mod, no generation
+		return;
+
+	if(!CResourceHandler::get("local")->createResource(filename))
+		return;
+	ResourcePath savePath(filename, EResType::IMAGE);
+
+	auto res = ImagePath::builtin("ADVOPTBK");
+	res = res.setOriginalResource(true);
+
+	std::shared_ptr<IImage> img = GH.renderHandler().loadImage(res, EImageBlitMode::OPAQUE);
+
+	Canvas canvas = Canvas(Point(575, 585), CanvasScalingPolicy::IGNORE);
+	canvas.draw(img, Point(0, 0), Rect(0, 0, 575, 585));
+	canvas.draw(img, Point(54, 121), Rect(54, 123, 335, 1));
+	canvas.draw(img, Point(158, 84), Rect(156, 84, 2, 37));
+	canvas.draw(img, Point(234, 84), Rect(232, 84, 2, 37));
+	canvas.draw(img, Point(310, 84), Rect(308, 84, 2, 37));
+	canvas.draw(img, Point(53, 567), Rect(53, 520, 339, 3));
+	canvas.draw(img, Point(53, 520), Rect(53, 264, 339, 47));
+
+	std::shared_ptr<IImage> image = GH.renderHandler().createImage(canvas.getInternalSurface());
+
+	image->exportBitmap(*CResourceHandler::get("local")->getResourceName(savePath));
+}
+
+void AssetGenerator::createBigSpellBook()
+{
+	std::string filename = "data/SpelBk2.bmp";
+
+	if(CResourceHandler::get()->existsResource(ResourcePath(filename))) // overridden by mod, no generation
+		return;
+
+	if(!CResourceHandler::get("local")->createResource(filename))
+		return;
+	ResourcePath savePath(filename, EResType::IMAGE);
+
+	auto res = ImagePath::builtin("SpelBack");
+	res = res.setOriginalResource(true);
```

Comment:
Perhaps instead of this flag we could create separate directory for autogenerated resources?
So if mod want to provide own spellbook image, vcmi will look for it in `data/SpellBookLarge.png`. But autogenerated image will be placed into `(local fs)/generated/data/SpellBookLarge.png`

Don't see why we need such flag as part of FS system.

Reply:
Yes, this is currently the case. assets are already generated in `(local fs)/generated/data/xyz.bmp`.

This flag is to force load the original, unmodded resource, even if there is a modded one. If someone will do a modded (**small**) spellbook: Without this flag the generation is done on the modded file. This may create a corrupted big spellbook (if there is no modded **big** spellbook).

Reply:
That sounds like a problem with mod. Mod should either provide valid small spellbook image or also provide large spellbook.

Sounds like an overkill solution for a problem that likely does not exists in the first place

Reply:
Hmm. Yes you're right. This could be a little bit overengineered.
Will remove.